libs/arangodb/langchain_arangodb/chains/graph_qa/arangodb.py

```diff
@@ -162,6 +172,187 @@ def from_llm(
         )
 
+    def _format_aql(self, aql: str) -> str:
+        keywords = [
+            "WITH",
+            "FOR",
+            "FILTER",
+            "LET",
+            "RETURN",
+            "SORT",
+            "LIMIT",
+            "COLLECT",
+            "INSERT",
+            "UPDATE",
+            "REMOVE",
+            "REPLACE",
+        ]
+        for kw in keywords:
+            # Add newline before keyword if it's not already preceded by one
+            aql = re.sub(rf"\b{kw}\b", f"\n{kw}", aql, flags=re.IGNORECASE)
+        return aql.strip()
+
+    def cache_query(
+        self, queries: Optional[List[Dict[str, str]]] = None
+    ) -> Union[str, List[str]]:
+        """
+        Cache one or more queries generated by the LLM only if
+        they're not already stored.
+
+        If no list is provided, caches the most recent query.
+
+        The query text is always lowercased before storing and checking for duplicates.
+
+        :param queries: Optional list of dicts with 'text' and 'aql' keys.
+        :return: True if a single new entry was inserted,
+                or list of messages for multiple entries.
+        """
+        if self.embedding is None:
+            raise ValueError("Cannot cache queries without an embedding model.")
+
+        collection = self.graph.db.collection(self.query_cache_collection_name)  # type: ignore
+
+        if queries is None:
+            # Fallback: cache the most recent query
+            text = self._last_user_input.strip().lower()  # type: ignore
+            aql = self._last_aql_query
+
+            existing = list(collection.find({"text": text, "aql": aql}, limit=1))
+            if existing:
+                return "This query is already in the cache"
+
+            query_embedding = self.embedding.embed_query(text)  # type: ignore
+            collection.insert(
+                {
+                    "text": text,
+                    "embedding": query_embedding,
+                    "aql": aql,
+                }
+            )  # type: ignore
+
+            return "Cached: " + self._last_user_input  # type: ignore
+
+        # Handle multiple queries
+        results = []
+        for q in queries:
+            text = q.get("text", "").strip().lower()
+            aql = q.get("aql", "")
+
+            existing = list(collection.find({"text": text}, limit=1))
+            if existing:
+                results.append(f"Skipped (already cached): {text}")
+                continue
+
+            embedding = self.embedding.embed_query(text)  # type: ignore
+            aql = self._format_aql(aql)
+
+            collection.insert(
+                {
+                    "text": text,
+                    "embedding": embedding,
+                    "aql": aql,
+                }
+            )  # type: ignore
+            results.append(f"Cached: {text}")
+
+        return results
+
+    def clear_query_cache(self, queries_to_remove: Optional[List[str]] = None) -> bool:
```

Comment:
same thing as `cache_query()`: let's just operate with one query at a time instead of a list of queries.

I.e 

```python

chain.clear_query_cache() # all queries removed

chain.clear_query_cache("Who is John Smith?")
```